enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/consensus/RaftMessages.java

```diff
@@ -1017,13 +1032,17 @@ public int hashCode()
 
     abstract class BaseRaftMessage implements RaftMessage
     {
-        protected MemberId from;
-        private Type type;
+        protected final MemberId from;
+        private final Type type;
+        private final Clock clock;
```

Comment:
I might be able to live with the recievedAt, but not the clock. Perhaps you can add another handler in the Netty-pipeline which fills in recievedAt?

Reply:
Keeping a reference to the clock makes it easier to reuse for creating messages in response. Alternatives aren't great:

* Don't inject a Clock, hard code instead
* Allow a null Instant, or make Optional
* Modify message handle methods to take a clock
* Make role handlers not singletons and inject clock
* Wrap messages in a timing aware wrapper similar to ClusterIdAwareMessage. Tried this, all of plain message / cluster ID message / timing message / cluster ID and timing message occur, it got messy. Could be the best solution, would be nicer in Scala :smiling_imp:

Reply:
I would be in favour of "Allow a null Instant, or make Optional" because this field is not really part of the message proper and should not be serialized. It is merely an additional tag on the message that gets added when it passes through the processing layers. Wrapping it would also be ok, but just having a special field is perhaps easier to work with and is also low overhead.